### PR TITLE
Add Python support

### PR DESCRIPTION
**Pull request by freitass**
Bringing Python support from original Dropbox-managed repository almost as-is and added some integration tests for it.

### CODE REVIEW

**Comment by a4z** [comment]
Since this is required for https://github.com/cross-language-cpp/djinni-support-lib/pull/24, I would say we pull it in and prepare a new release. I have some success on my local mach with that, for both,  Linux and Mac.

I think these PRs should be done in a way leading us to 1.0.0, 
maybe a 0.99.0 , and once everything is fine we change both , generator and support lib , to 1.0.0 .. ?

**Comment by jothepro** [comment]
We haven't talked about a versioning concept in djinni yet but I think we should stick to [semantic versioning](https://semver.org/). I would be fine with making an exception to send a signal that "we are feature complete now", but personally I'd like to wait with that until C# support is also in master.

**Comment by jothepro** [comment]
First of all: Super cool, @freitass! Thanks for your contribution!!!

I am missing documentation on about how the djinni integration works:

- "[Interface Definition Language (IDL)](https://djinni.xlcpp.dev/djinni-generator/idl/)":
    - Updated example listings to include `+p` modifier
    - Datatypes: How do C++ types map to Python types (date, map, ...)?
    - Exceptions: How does the Python integration handle Exceptions?
- "[Using the generator](https://djinni.xlcpp.dev/djinni-generator/cli-usage/)": Documentation of existing command line parameters (basically the same information that is output by `djinni --help`.
- "[Use generated Code](https://djinni.xlcpp.dev/djinni-generator/generated-code-usage/)": 
    - What files are generated?
    - A bit of background how the Python <-> C++ interaction works. I am aware that this is new for us all so I think it cannot be super in-depth, but a few sentences would help me and hopefully others to understand what is going on.

You can find the Markdown-files that generate the documentation in `docs`.

File: src/main/scala/djinni/Main.scala
Comment (jothepro) [inline src/main/scala/djinni/Main.scala:237]:
```suggestion
        .text("The import prefix used within python generated files (default: \"\")")
```

**Review by jothepro — COMMENTED** [review]
Wow... I am migrating the C# support in parallel and I have a feeling like the python integration is so much more complex... 😮 At least it generates a lot more gluecode 🙈 .
At a first glance the only missing feature in the generator seems to be the ability to set a header import prefix for the support-lib (equivalent to `--jni-base-lib-include-prefix`. Do you plan to add it?

File: src/it/resources/expected/my_flags/python/my_flags.py
Comment (freitass) [reply src/it/resources/expected/my_flags/python/my_flags.py:17]: It should be investigated.

**Comment by freitass** [comment]
> At a first glance the only missing feature in the generator seems to be the ability to set a header import prefix for the support-lib (equivalent to `--jni-base-lib-include-prefix`. Do you plan to add it?

We don't have any new features planned for Djinni, but it'd be cool to have that option to make it similar to the other languages!

**Comment by freitass** [comment]
> I am migrating the C# support in parallel and I have a feeling like the python integration is so much more complex... 😮

Indeed! It'll be much harder to document Python in https://djinni.xlcpp.dev/djinni-generator/generated-code-usage/. I believe that's our only big piece of documentation missing right now.

File: src/it/resources/expected/my_flags/python/my_flags.py
Comment (jothepro) [reply src/it/resources/expected/my_flags/python/my_flags.py:17]: Maybe because `IntFlag` was only introduced in Python 3.6 ([link](https://docs.python.org/3/library/enum.html#enum.auto)), which was released in late 2016. Do you have a minimum python3 version in mind, that should be supported?

Comment (a4z) [reply src/it/resources/expected/my_flags/python/my_flags.py:17]: 3.6 should be fine, earlier versions are not supported anymore, and we need to think to the future, not the past. 
3.6 will end this year, with supporting that version  we would be as much backward compatible as it makes sense.

**Comment by paulo-coutinho** [comment]
Hi.

First, good job man, very thanks for that. Im reading the files changes to review and have two questions:

1. The cwrapper output files can be work with other things like flutter or is specific for python?
2. The yaml files can be used with python generator too?

Thanks.

File: src/it/resources/expected/my_flags/python/my_flags.py
Comment (freitass) [reply src/it/resources/expected/my_flags/python/my_flags.py]: I just pushed a change to use `IntFlag` for Djinni flags in Python. Resulting code looks like this: https://github.com/freitass/djinni-generator/blob/93c9e4cbb952ae17558123d5b5b2515a2a152da1/src/it/resources/expected/my_flags/python/my_flags.py#L9-L16

File: docs/generated-code-usage.md
Comment (jothepro) [inline docs/generated-code-usage.md:128]: I've introduced

- `--c-wrapper-include-prefix <prefix>`
- `--c-wrapper-include-cpp-prefix <prefix>`
- `--c-wrapper-base-lib-include-prefix <prefix>`

So from my point of view header prefixes are as configurable as they can get. Am I missing something?

File: src/it/resources/expected/my_enum/python/my_enum.py
Comment (jothepro) [inline src/it/resources/expected/my_enum/python/my_enum.py]: Is there a reason why the `@unique` annotation had to be removed? Is it incompatible with `IntFlag`?

File: docs/generated-code-usage.md
Comment (freitass) [reply docs/generated-code-usage.md:128]: You're right. I didn't realize how many of the limitations we've already addressed so far!

File: src/it/resources/expected/my_enum/python/my_enum.py
Comment (freitass) [reply src/it/resources/expected/my_enum/python/my_enum.py]: I concluded [from the docs](https://docs.python.org/3/library/enum.html) it only applies to `Enum` and since the code is generated anyways I didn't see the value of having it (even if it was applicable).

**Comment by freitass** [comment]
@jothepro I believe we've improved all points mentioned in [this comment](https://github.com/cross-language-cpp/djinni-generator/pull/43#issuecomment-818498894), can you confirm?

**Review by jothepro — APPROVED** [review]
Awesome!!! Let's bring this to master! 🚀  Thank you for your passionate contribution, @freitass!

I did a final review and these are the last two things that caught my eyes:

- Do you know wether record [extension](https://github.com/freitass/djinni-generator/blob/python/docs/idl.md#extensions) is supported in Python? It's not mentioned in the docs.
- I found some more autogenerated headers that were still missing a "was". I've made a [PR for this](https://github.com/freitass/djinni-generator/pull/3).

I have not yet tested the generated code on my own! My review is based on wether the generators output looks reasonable to me. I trust you that you have successfully used the Python integration at least in a simple test setup once!
I am ok with the risk that I do not at all understand the details of this implementation. We'll learn about those details when things break 😉 .

**Review by a4z — APPROVED** [review]
Guys, you are doing an awesome job! 

Thanks a lot!!

**Comment by a4z** [comment]
and I agree, let's bring that to master, we then can enable uint tests for the support lib and start iterating over the states of both.
maybe we will also get additional testers when its there to use ....

**Comment by jothepro** [comment]
👍 ok, @freitass I'll wait for you to merge my PR and answer my question. Once I got your response I will merge into master and make a release. Then we can start focusing on the support-lib.

**Comment by a4z** [comment]
what we would need would be something like a latest generator build, to download, because now the support lib still doesn't have any way to get it (accept, getting the code and building it at every PR, what is not optimal) 
@jothepro , do you see an easy way to always have a latest to download after a merge into master?

**Comment by jothepro** [comment]
I don't see a problem there (yet). I think it is a good design decision to make the support-lib depend on a specified release version instead of an undefined 'latest' state of the generator.
No, I do not have a decent idea for that right now.

**Comment by a4z** [comment]
this means, we need a release now to run the unit tests .... 
and if we (would) need one or 2 fast iterations for fixing things in the generator/support lib , this would mean one or 2 fast releases for nothing ... I think  I need to think about that 
.... latests would not be undefined, it's the latest .... maybe attaching it as artefact to a gh action , we will see

**Comment by freitass** [comment]
> * Do you know wether record [extension](https://github.com/freitass/djinni-generator/blob/python/docs/idl.md#extensions) is supported in Python? It's not mentioned in the docs.
> * I found some more autogenerated headers that were still missing a "was". I've made a [PR for this](https://github.com/freitass/djinni-generator/pull/3).

Check and check! ✅

**Pull request merged**
